### PR TITLE
add optional feature to check modified file modified time to return 304

### PR DESCRIPTION
**Pull request by avimar**
https://github.com/restify/node-restify/pull/994

https://github.com/restify/plugins/pull/17
@Dzitu what version of windows return milliseconds? My Windows 7 always ends with `000` meaning it's only accurate to seconds.

### CODE REVIEW

**Comment by avimar** [comment]
(Just to note, this effectively ignores bad if-modified-since headers, it won't trigger an error.)

**Comment by Dzitu** [comment]
@avimar Windows 7 here as well, following script:
`var fs = require("fs");`
`var path = "static.js";`
`fs.stat(path, function(err, stats) {`
`if(!err)`
`console.log(new Date(stats.mtime).getMilliseconds());`
`});`

Returns 491 for me.

**Comment by avimar** [comment]
@Dzitu Interesting!
Turns out it not windows or linux, but rather... `Date.parse()` strips milliseconds.

I don't know if that will ever change..
but since `conditional` and my implementation use `Date.parse()` it all seems OK.

``` javascript
var fs=require('fs');
var stat = fs.statSync('myFile');
console.log(new Date(stat.mtime).getMilliseconds());
//769
console.log(new Date(stat.mtime).valueOf())
//1449737820769
console.log(Date.parse(stat.mtime))
//1449737820000 <-- no milliseconds
```

**Comment by DonutEspresso** [comment]
LGTM, @micahr or @yunong to verify please.

File: lib/plugins/static.js [hunk 1]
```diff
@@ -55,29 +56,48 @@ function serveStatic(options) {
...
-            }
+        res.cache({maxAge: maxAge});
+        res.set('Content-Length', stats.size);
+        res.set('Content-Type', mime.lookup(file));
+        res.set('Last-Modified', stats.mtime);
+
+        if (opts.charSet) {
+            var type = res.getHeader('Content-Type') +
+                '; charset=' + opts.charSet;
+            res.setHeader('Content-Type', type);
```
Comment (micahr) [inline lib/plugins/static.js:68]: Will this override the `res.set('Content-Type')` from before? If so, perhaps they should be using the same method?

File: lib/plugins/static.js [hunk 2]
```diff
@@ -55,29 +56,48 @@ function serveStatic(options) {
...
+            if (opts.checkIfModified) {
+                requestTime = Date.parse(req.header('if-modified-since')) || 0;
+                modifiedTime = Date.parse(stats.mtime);
             }
-            res.writeHead(200);
-            fstream.pipe(res);
-            fstream.once('end', function () {
-                next(false);
+        } catch (e) {
+            opts.checkIfModified = false;
```
Comment (micahr) [inline lib/plugins/static.js:84]: This error shouldn't be swallowed. What is the expected error? I believe `Date.parse` doesn't throw an error like JSON.parse, rather it returns NaN if it encounters invalid data. This try catch shouldn't be necessary.

**Comment by micahr** [comment]
I added a few comments. I'd love to see a test or two for this change as well.

File: lib/plugins/static.js (hunk 2 above)
Comment (avimar) [inline lib/plugins/static.js:84]: There's a v8 bug that causes Date.parse to throw: https://github.com/jfhbrook/node-ecstatic/pull/179
I was able to replicate this in node 4.3.0: https://github.com/restify/node-restify/pull/994#issuecomment-198041267

File: lib/plugins/static.js (hunk 1 above)
Comment (avimar) [inline lib/plugins/static.js:68]: I don't know: I only moved that code, I didn't write it. It looks like it appends it to the current content-type.

**Comment by avimar** [comment]
@micahr is tests holding this up?
I looked at the test suite and I don't really understand it. I'd assume the same test that applies to the if-conditions can be re-applied with a temp file and running `static` in the middle.

**Comment by avimar** [comment]
@micahr - I addressed your comments. What now?

**Comment by yunong** [comment]
@avimar the only thing left now is to add some unit tests for the new behaviour.

**Comment by retrohacker** [comment]
Hey @avimar,

It looks like @yunong requested tests a while back, though you were caught up on the testing framework and requested assistance. We are in the process of moving the plugins back into the main repo. If this work is still important to you, please re-open a PR over on that repo and I would be more than guide you through putting together some tests :heart:

**Pull request closed without merging**